// regraph-next/build/ReScreen/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function isValidSVG(parentNode) {
    if (!parentNode) {
        console.log("svg元素不存在");
        return false;
    }
    var node = parentNode.node();
    var childNodes = node.childNodes;
    if (!childNodes) {
        return false;
    }
    else if (childNodes.length > 1) {
        return false;
    }
    else {
        var nodeName = childNodes[0].nodeName;
        return nodeName === "g";
    }
}
exports.isValidSVG = isValidSVG;
function requestFullscreen(element) {
    if (element.requestFullscreen) {
        element.requestFullscreen();
    }
    else if (element.mozRequestFullScreen) {
        element.mozRequestFullScreen();
    }
    else if (element.msRequestFullscreen) {
        element.msRequestFullscreen();
    }
    else if (element.webkitRequestFullscreen) {
        element.webkitRequestFullscreen();
    }
}
exports.requestFullscreen = requestFullscreen;
function exitFullscreen() {
    if (document.exitFullscreen) {
        document.exitFullscreen();
    }
    else if (document.exitFullscreen) {
        document.exitFullscreen();
    }
    else if (document.webkitExitFullscreen) {
        document.webkitExitFullscreen();
    }
}
exports.exitFullscreen = exitFullscreen;
function getContainerStyle(mapPositon) {
    switch (mapPositon) {
        case "RB":
            return {
                alignItems: "flex-end"
            };
        case "LT":
            return {
                flexDirection: "row-reverse"
            };
        case "LB":
            return {
                alignItems: "flex-end",
                flexDirection: "row-reverse"
            };
        default:
            return;
    }
}
exports.getContainerStyle = getContainerStyle;
function getMapStyle(mapPositon) {
    switch (mapPositon) {
        case "RB-IN":
            return {
                position: "absolute",
                right: 0,
                bottom: 0
            };
        case "LT-IN":
            return {
                position: "absolute",
                left: 0,
                top: 0
            };
        case "LB-IN":
            return {
                position: "absolute",
                left: 0,
                bottom: 0
            };
        case "RT-IN":
            return {
                position: "absolute",
                right: 0,
                top: 0
            };
        default:
            return;
    }
}
exports.getMapStyle = getMapStyle;
